### PR TITLE
Use byte[] keys in FormatWriter to save on hashmap lookups when the keys are constant

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/JsonFormatWriter.java [hunk 1]
```diff
@@ -23,8 +23,8 @@
   public static JsonFormatWriter JSON_WRITER = new JsonFormatWriter();
 
   @Override
-  public void writeKey(final String key, final JsonWriter destination) throws IOException {
-    destination.name(key);
+  public void writeKey(byte[] key, JsonWriter destination) throws IOException {
+    destination.name(new String(key));
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/JsonFormatWriter.java:27]: I guess JsonFormatWriter is not performance critical, so the extra `String` allocations here are not a concern.

Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/core/serialization/JsonFormatWriter.java:27]: Exactly, but I want to do a bit more due diligence on this to see if it's worth the drop in code quality.

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java [hunk 2]
```diff
@@ -51,6 +52,59 @@ public void writeTag(String key, String value, MessagePacker destination) throws
     }
   }
 
+  @Override
+  public void writeString(byte[] key, String value, MessagePacker destination) throws IOException {
+    writeKey(key, destination);
+    destination.packString(value);
```
Comment (lpriima) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java:58]: NPE happens here, when DDSpanContext#getSpanType() returns `null`


compare to : https://github.com/DataDog/dd-trace-java/pull/1455/files#diff-d27443624dce2e3e0340de2124be31aeR37

Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java]: addressed, would you be willing to approve please?

**Pull request merged**